src/encoder.rs

```diff
@@ -1232,39 +1236,63 @@ pub fn encode_tx_block(
   tx_size: TxSize, tx_type: TxType, plane_bsize: BlockSize, po: &PlaneOffset,
-  skip: bool, bit_depth: usize, ac: &[i16], alpha: i16
-) -> bool {
+  skip: bool, bit_depth: usize, ac: &[i16], alpha: i16, for_rdo_use: bool
+) -> (bool, i64) {
     let rec = &mut fs.rec.planes[p];
     let PlaneConfig { stride, xdec, ydec, .. } = fs.input.planes[p].cfg;
 
     if mode.is_intra() {
       mode.predict_intra(&mut rec.mut_slice(po), tx_size, bit_depth, &ac, alpha);
     }
 
-    if skip { return false; }
+    if skip { return (false, 0); }
 
-    let mut residual: AlignedArray<[i16; 64 * 64]> = UninitializedAlignedArray();
+    let mut residual_storage: AlignedArray<[i16; 64 * 64]> = UninitializedAlignedArray();
     let mut coeffs_storage: AlignedArray<[i32; 64 * 64]> = UninitializedAlignedArray();
-    let mut rcoeffs: AlignedArray<[i32; 64 * 64]> = UninitializedAlignedArray();
+    let mut qcoeffs_storage: AlignedArray<[i32; 64 * 64]> = UninitializedAlignedArray();
+    let mut rcoeffs_storage: AlignedArray<[i32; 64 * 64]> = UninitializedAlignedArray();
+    let residual = &mut residual_storage.array[..tx_size.area()];
     let coeffs = &mut coeffs_storage.array[..tx_size.area()];
+    let qcoeffs = &mut qcoeffs_storage.array[..tx_size.area()];
+    let rcoeffs = &mut rcoeffs_storage.array[..tx_size.area()];
 
-    diff(&mut residual.array,
+    diff(residual,
          &fs.input.planes[p].slice(po),
          &rec.slice(po),
          tx_size.width(),
          tx_size.height());
 
-    forward_transform(&residual.array, coeffs, tx_size.width(), tx_size, tx_type, bit_depth);
-    fs.qc.quantize(coeffs);
+    forward_transform(residual, coeffs, tx_size.width(), tx_size, tx_type, bit_depth);
+    fs.qc.quantize(coeffs, qcoeffs);
 
-    let has_coeff = cw.write_coeffs_lv_map(w, p, bo, &coeffs, mode, tx_size, tx_type, plane_bsize, xdec, ydec,
+    let has_coeff = cw.write_coeffs_lv_map(w, p, bo, &qcoeffs, mode, tx_size, tx_type, plane_bsize, xdec, ydec,
                             fi.use_reduced_tx_set);
 
     // Reconstruct
-    dequantize(fi.base_q_idx, &coeffs, &mut rcoeffs.array, tx_size, bit_depth);
+    dequantize(fi.base_q_idx, qcoeffs, rcoeffs, tx_size, bit_depth);
+
+    let mut tx_dist: i64 = -1;
 
-    inverse_transform_add(&rcoeffs.array, &mut rec.mut_slice(po).as_mut_slice(), stride, tx_size, tx_type, bit_depth);
-    has_coeff
+    if !fi.use_tx_domain_distortion || !for_rdo_use {
+    //if !fi.use_tx_domain_distortion {
+        inverse_transform_add(rcoeffs, &mut rec.mut_slice(po).as_mut_slice(), stride, tx_size, tx_type, bit_depth);
+    } else {
+        //inverse_transform_add(rcoeffs, &mut rec.mut_slice(po).as_mut_slice(), stride, tx_size, tx_type, bit_depth);
+
+        // Store tx-domain distortion of this block
+        tx_dist = coeffs
+            .iter()
+            .zip(rcoeffs)
+            .map(|(a, b)| {
+                let c = (*a as i16 - *b as i16) as i32;
```

Comment:
You are right, each tx coeff with 9 bits residue input can need full 15 bits + 1 bit (sign), so add/subtract two of them need one more bit to store the result, correct?
Furthermore, for the 10/12 bits input, I think it will need more precision than 16  + 1 bits.